File: android/app/src/main/java/com/nearpair/app/model/TransferModels.kt
package com.nearpair.app.model

import java.io.File

data class NearbyDevice(
    val endpointId: String,
    val name: String,
)

data class StagedFile(
    val file: File,
    val displayName: String,
    val mimeType: String,
    val sizeBytes: Long,
    val sha256: String,
)

data class ReceivedFile(
    val file: File,
    val displayName: String,
    val mimeType: String,
    val sizeBytes: Long,
    val sha256: String,
)

enum class TransferDirection { SEND, RECEIVE }

enum class FailureReason {
    PERMISSION_DENIED,
    APP_CONFIGURATION,
    RADIOS_DISABLED,
    NO_DEVICES,
    REJECTED,
    CONNECTION_LOST,
    INSUFFICIENT_STORAGE,
    INVALID_METADATA,
    UNSUPPORTED_TYPE,
    SIZE_MISMATCH,
    CHECKSUM_MISMATCH,
    IO_FAILURE,
    PROTOCOL_MISMATCH,
    BACKGROUNDED,
    UNKNOWN,
}

sealed interface TransferState {
    data object Idle : TransferState

    data class PermissionsNeeded(val action: String) : TransferState

    data class Advertising(val deviceName: String) : TransferState

    data class Discovering(val devices: List<NearbyDevice>) : TransferState

    data class ConnectionRequested(val device: NearbyDevice) : TransferState

    data class ConfirmCode(
        val endpointId: String,
        val deviceName: String,
        val code: String,
    ) : TransferState

    data class Connected(
        val endpointId: String,
        val deviceName: String,
    ) : TransferState

    data class Transferring(
        val direction: TransferDirection,
        val fileName: String,
        val transferredBytes: Long,
        val totalBytes: Long,
    ) : TransferState {
        val progress: Float
            get() = if (totalBytes <= 0L) 0f else (transferredBytes.toDouble() / totalBytes).toFloat()
                .coerceIn(0f, 1f)
    }

    data class Verifying(
        val direction: TransferDirection,
        val fileName: String,
    ) : TransferState

    data class Complete(
        val direction: TransferDirection,
        val fileName: String,
    ) : TransferState

    data class Failed(
        val reason: FailureReason,
        val detail: String,
    ) : TransferState

    data object Cancelled : TransferState
}

File: ios/NearPair/Models/TransferModels.swift
import Foundation

struct NearbyDevice: Identifiable, Equatable {
    let id: String
    let name: String
}

struct StagedFile: Equatable {
    let url: URL
    let displayName: String
    let mimeType: String
    let sizeBytes: Int64
    let sha256: String
}

struct ReceivedFile: Identifiable, Equatable {
    var id: URL { url }
    let url: URL
    let displayName: String
    let mimeType: String
    let sizeBytes: Int64
    let sha256: String
}

enum TransferDirection: Equatable {
    case send
    case receive
}

enum FailureReason: Equatable {
    case permissionDenied
    case radiosDisabled
    case rejected
    case connectionLost
    case insufficientStorage
    case invalidMetadata
    case unsupportedType
    case sizeMismatch
    case checksumMismatch
    case ioFailure
    case protocolMismatch
    case backgrounded
    case unknown
}

enum TransferPhase: Equatable {
    case idle
    case advertising(deviceName: String)
    case discovering
    case connectionRequested(device: NearbyDevice)
    case confirmCode(endpointID: String, deviceName: String, code: String)
    case connected(endpointID: String, deviceName: String)
    case transferring(direction: TransferDirection, fileName: String, transferredBytes: Int64, totalBytes: Int64)
    case verifying(direction: TransferDirection, fileName: String)
    case complete(direction: TransferDirection, fileName: String)
    case failed(reason: FailureReason, detail: String)
    case cancelled
}

@MainActor
protocol TransferEngine: AnyObject {
    var phase: TransferPhase { get }
    var devices: [NearbyDevice] { get }
    var receivedFile: ReceivedFile? { get }

    func startReceiving(deviceName: String)
    func stopReceiving()
    func discoverNearbyDevices(deviceName: String)
    func stopDiscovery()
    func requestConnection(to device: NearbyDevice, localDeviceName: String)
    func acceptConnection(endpointID: String)
    func rejectConnection(endpointID: String)
    func sendFile(_ file: StagedFile)
    func cancelTransfer()
    func reset()
    func onAppBackgrounded()
}

